### PR TITLE
T204106: Add classes PerFactException and PerFactUserWarning

### PR DESCRIPTION
**Pull request by hvanaart**
Add two classes "PerFactException" and "PerFactUserWarning" to enable dynamic error message formating.
That enables translation to be applied to the text template before rendering the final message.

### CODE REVIEW

**Comment by jan-jockusch** [comment]
@viktordick This has been implemented according to your guidance, so I've been told. The motivation for the new exception class is clear, and the execution seems to be clean. What I need to know: Does this match your specifications?

**Comment by hvanaart** [comment]
@jan-jockusch  @viktordick The only thing kept secret from Viktor is the overloading of the __str__ method in PerFactException. This was necessary for writing the error message to the standard output, when the exception is raised. By default the error message is error_value which is a string. In case of the custom class, the error_value is the error object. Without customization no value would be printed, as the object could not be casted into a string. In this case a set of the template text with placeholders + the dynamic values are used for printout.

File: Products/PerFactErrors/errors.py [hunk 1]
```diff
@@ -50,7 +51,53 @@ def afterfail_error_message(event):
         if body is not None:
             req.response.setBody(body)
     except Exception:
+        logger.warn(traceback.format_exc())
```
Comment (viktordick) [inline Products/PerFactErrors/errors.py]: Why is this here? `logger.exception` in the next line already includes the current traceback.

File: Products/PerFactErrors/errors.py [hunk 2]
```diff
@@ -50,7 +51,53 @@ def afterfail_error_message(event):
...
+        '''
+
+        super(PerFactException, self).__init__()
+        self.msg = msg
+        self.show_to_user = show_to_user
+        self.apperrorlog = apperrorlog
+        self.payload = payload or {}
+
+    def __str__(self):
+        return repr((self.msg, self.payload))
```
Comment (viktordick) [inline Products/PerFactErrors/errors.py:97]: Sorry I did not understand why this is necessary. The error handling should not stringify instances of `PerFactException`. But I guess I will see that when reviewing the changes in the Data.FS. By itself and for debugging this does not hurt.

File: Products/PerFactErrors/errors.py [hunk 3]
```diff
@@ -50,7 +51,53 @@ def afterfail_error_message(event):
...
+
+
+class PerFactUserWarning(PerFactException):
+
```
Comment (viktordick) [inline Products/PerFactErrors/errors.py:107]: Please add a comment about how this class differs from its base.

File: Products/PerFactErrors/errors.py [hunk 4]
```diff
@@ -50,7 +51,53 @@ def afterfail_error_message(event):
...
     except Exception:
+        logger.warn(traceback.format_exc())
         logger.exception('Error while rendering error message')
         transaction.abort()
     else:  # no exception
         transaction.commit()
+
+
+class PerFactException(Exception):
+    ''' Special Exception Class
```
Comment (viktordick) [inline Products/PerFactErrors/errors.py]: A sentence here would be nice explaining the purpose of the class. Most of it can be deduced from the docstring of `__init__`, but anyhow.

**Review by viktordick — CHANGES_REQUESTED** [review]
Just some small adjustments needed.

File: Products/PerFactErrors/errors.py (hunk 2 above)
Comment (hvanaart) [reply Products/PerFactErrors/errors.py:97]: Without the function overloading, no error message was printed. The only information you got in the log was the exception type.

File: Products/PerFactErrors/errors.py (hunk 1 above)
Comment (hvanaart) [reply Products/PerFactErrors/errors.py]: I checked the traceback in the result again. I will drop this line

**Review by viktordick — APPROVED** [review]
LGTM
